libraries/AP_AIS/AP_AIS.cpp

```diff
@@ -0,0 +1,764 @@
+    AP_GROUPINFO_FLAGS("TYPE", 1, AP_AIS, _type, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: LIST_MAX
+    // @DisplayName: AIS vessel list size
+    // @Description: AIS list size of nearest vessels. Longer lists take longer to refresh with lower SRx_ADSB values.
+    // @Range: 1 100
+    // @User: Advanced
+    AP_GROUPINFO("LIST_MAX", 2, AP_AIS, _max_list, LIST_SIZE_DEFAULT),
+
+    // @Param: TIME_OUT
+    // @DisplayName: AIS vessel time out
+    // @Description: if no updates are received in this time a vessel will be removed from the list
+    // @Units: s
+    // @Range: 1 2000
+    // @User: Advanced
+    AP_GROUPINFO("TIME_OUT", 3, AP_AIS, _time_out, TIME_OUT_DEFAULT),
+
+    // @Param: OPTIONS
+    // @DisplayName: AIS options
+    // @Description: Bitmask of AIS options
+    // @Bitmask: 0:Log all AIVDM messages,1:Log only unsupported AIVDM messages,2:Log decoded messages
+    // @User: Advanced
+    AP_GROUPINFO("OPTIONS", 4, AP_AIS, _options, AIS_OPTIONS_LOG_UNSUPPORTED_RAW),
+
+    AP_GROUPEND
+};
+
+// constructor
+AP_AIS::AP_AIS() :
+    _list(8)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+// Initialize the AIS object and prepare it for use
+void AP_AIS::init()
+{
+    if (!enabled()) {
+        return;
+    }
+
+    _uart = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_AIS, 0);
+    if (_uart == nullptr) {
+        return;
+    }
+
+    _uart->begin(AP::serialmanager().find_baudrate(AP_SerialManager::SerialProtocol_AIS, 0));
+}
+
+// update AIS, expected to be called at 20hz
+void AP_AIS::update()
+{
+    if (!_uart || !enabled()) {
+        return;
+    }
+
+    // read any available lines
+    u_int32_t nbytes = MAX(_uart->available(),1024U);
+    while (nbytes-- > 0) {
+        const int16_t byte = _uart->read();
+        if (byte == -1) {
+            break;
+        }
+        const char c = byte;
+        if (decode(c)) {
+            const bool log_all = (_options & AIS_OPTIONS_LOG_ALL_RAW) != 0;
+            const bool log_unsupported = ((_options & AIS_OPTIONS_LOG_UNSUPPORTED_RAW) != 0) && !log_all; // only log unsupported if not logging all
+
+            if (_incoming.total  > AIVDM_BUFFER_SIZE)  {
+                // no point in trying to decode it wont fit
+                if (log_all || log_unsupported) {
+                    log_raw(&_incoming);
+                }
+                break;
+            }
+            if (log_all) {
+                log_raw(&_incoming);
+            }
+
+            if (_incoming.num == 1 && _incoming.total == 1) {
+                // single part message
+                if (!payload_decode(_incoming.payload) && log_unsupported) {
+                    // could not decode so log
+                    log_raw(&_incoming);
+                }
+            } else if (_incoming.num == _incoming.total) {
+                // last part of a multi part message
+                uint8_t index = 0;
+                uint8_t msg_parts[_incoming.total - 1];
```

Comment:
```suggestion
                uint8_t msg_parts[_incoming.num - 1];
```

These are equal - but the bounds-checking for index is based on `num`, not `total`.